docs/madr/decisions/014-mesh-circuit-breaker-policy.md

```diff
@@ -0,0 +1,198 @@
+# Circuit Breaker policy compliant with 2.0 model
+
+* Status: accepted
+
+Technical Story: [#4736](https://github.com/kumahq/kuma/issues/4736)
+
+## Context and Problem Statement
+
+[New policy matching MADR](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md)
+introduces a new approach how Kuma applies configuration to proxies. Rolling out strategy implies creating a new set of
+policies that satisfy a new policy matching. Current MADR aims to define a timeout policy that will be compliant with
+2.0 policies model.
+
+Current `CircuitBreaker` policy combines
+envoy's [circuit breaking](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking)
+and [outlier detection](https://gstwww.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier)
+capabilities behind the scene, but there is no clear differentiation between two in the configuration specification. It
+may be misleading that we have a policy named the same as some envoy functionality, but it actually combines more
+functionalities. Consequences of this approach may introduce the situation when person more familiar with envoy during
+debugging process, may look only at stats prefixed with `cluster.<name>.circuit_breakers.`, but according to the
+behaviour of our policy, if the data plane was ejected because of one of the faults defined in `detectors` section,
+stats which contain interesting us data will be prefixed with `cluster.<name>.outlier_detection.`.
+
+## Considered Options
+
+* Create two separate policies: `MeshCircuitBreaker` and `MeshOutlierDetector`
+* Create single `MeshCircuitBreaker` policy which would map 1:1 current policy
+* Create single `MeshCircuitBreaker` with configuration divided
+  to `connectionPool` ([envoy's circuit breaker](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking))
+  and `outlierDetection` ([envoy's outlier detection](https://gstwww.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier))
+
+## Decision Outcome
+
+Chosen option: Create single `MeshCircuitBreaker` with configuration divided into two sections.
```

Comment:
There is one big advantage of separate policies that I see, we could finally apply this also on inbound side
```
type: MeshConnectionLimits
mesh: default
name: default-limits
spec:
  targetRef:
    kind: Mesh
    name: default
  from:
  - targetRef:
      kind: Mesh # only Mesh is available
      name: default
    default:
      connectionPool:
        maxConnections: 1024
        maxPendingRequests: 1024
        maxRetries: 3
        maxRequests: 1024
  to:
  - targetRef:
      kind: Mesh
      name: default
    default:
      connectionPool:
        maxConnections: 1024
        maxPendingRequests: 1024
        maxRetries: 3
        maxRequests: 1024
```

edit: ok, technically we don't need to split policy for this. We could also set outlier detection on inbound clusters.

Whether we split or not, it would be nice to finally be able to configure this on inbound.

Reply:
Hmmm, that's a good point

Reply:
yes, that's the point.